### PR TITLE
Implement f128 @rem

### LINKED ISSUE #10819: stage1: @rem f128 incorrect value
### Zig Version

0.9.0, 0.10.0-dev.490+dd7309bde

### Steps to Reproduce

Example in Zig:

```zig
const builtin = @import("builtin");
const std = @import("std");

pub fn main() void {
    const a = 6.9;
    const b = 4.0;
    const rem_f64 = remdiv_f64(a, b) catch unreachable;
    const rem_f128 = remdiv_f128(a, b) catch unreachable;
    const ct_rem_f64 = comptime try remdiv_f64(a, b);
    const ct_rem_f128 = comptime try remdiv_f128(a, b);

    std.debug.print("[f64]  @rem({}, {}) = {}\n", .{ a, b, rem_f64 });
    std.debug.print("[f128] @rem({}, {}) = {}\n", .{ a, b, rem_f128 });
    std.debug.print("[f64]  comptime @rem({}, {}) = {}\n", .{ a, b, ct_rem_f64 });
    std.debug.print("[f128] comptime @rem({}, {}) = {}\n", .{ a, b, ct_rem_f128 });
}

fn remdiv_f128(x: f128, y: f128) !f128 {
    return @rem(x, y);
}

fn remdiv_f64(x: f64, y: f64) !f64 {
    return @rem(x, y);
}
```

Build with:
zig build-exe file.zig -fLLVM -lm

### Expected Behavior

Each call should produce `2.9`, expected terminal output:
```
[f64]  @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
[f128] @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
[f64]  comptime @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
[f128] comptime @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
```

### Actual Behavior

```
[f64]  @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
[f128] @rem(6.9e+00, 4.0e+00) = 6.9e+00
[f64]  comptime @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
[f128] comptime @rem(6.9e+00, 4.0e+00) = -1.1e+00
```

### PR DESCRIPTION
**Pull request by m-radomski**
This contains code from #10827. Also as suggested by @matu3ba, added regression tests.

Presented example code in the closing issue now results in
```
[f16]  @rem(6.9e+00, 4.0e+00) = 2.8984375e+00
[f64]  @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
[f80]  @rem(6.9e+00, 4.0e+00) = -nan
[f128] @rem(6.9e+00, 4.0e+00) = 2.9e+00
[f16]  comptime @rem(6.9e+00, 4.0e+00) = 2.8984375e+00
[f64]  comptime @rem(6.9e+00, 4.0e+00) = 2.9000000000000004e+00
[f80]  comptime @rem(6.9e+00, 4.0e+00) = 2.9e+00
[f128] comptime @rem(6.9e+00, 4.0e+00) = 2.9e+00
```

Closes #10819.

### CODE REVIEW

File: lib/std/special/compiler_rt/floatfmodl_test.zig [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+const fmodl = @import("floatfmodl.zig");
+const testing = std.testing;
+
+fn test_fmodl(a: f128, b: f128, exp: f128) !void {
+    const res = fmodl.fmodl(a, b);
+    try testing.expect(exp == res);
+}
+
+test "fmodl" {
+    try test_fmodl(6.8, 4.0, 2.8);
```
Comment (matu3ba) [inline lib/std/special/compiler_rt/floatfmodl_test.zig:11]: what about the edge cases of float stuff, ie Nan, +-inf, denormalized numbers, +-0 ?
are they handled?
adding 2 negative numbers for the mod also would be helpful.

if you want to go the long road, you can add one or two bigger numbers.

File: lib/std/special/compiler_rt/floatfmodl.zig [hunk 2]
```diff
@@ -0,0 +1,129 @@
+const builtin = @import("builtin");
+const std = @import("std");
+
+// fmodl - floating modulo large, returns the remainder of division for f128 types
+// Logic and flow heavily inspired by MUSL fmodl for 113 mantissa digits
+pub fn fmodl(a: f128, b: f128) callconv(.C) f128 {
+    @setRuntimeSafety(false);
```
Comment (matu3ba) [inline lib/std/special/compiler_rt/floatfmodl.zig:7]: You dont have a big amount of tests, so why not `@setRuntimeSafety(builtin.is_test);` ?

File: src/stage1/ir.cpp [hunk 3]
```diff
@@ -3338,6 +3338,32 @@ static void float_div_floor(ZigValue *out_val, ZigValue *op1, ZigValue *op2) {
     }
 }
 
+// c = a - b * trunc(a / b)
+static float16_t zig_f16_rem(float16_t a, float16_t b) {
```
Comment (matu3ba) [inline src/stage1/ir.cpp:3342]: why is this called `rem`, if before it was called `mod` ?
[modulo != remainder](https://stackoverflow.com/a/13683709) for negative numbers.
Add test cases with negative numbers.

**Review by matu3ba — CHANGES_REQUESTED** [review]
`floatmod` looks nice and cozy, but use it for `rem` without explaining how one uses the other in the commit message.

mod and rem are different for negative numbers.
add a few more tests, especially negative numbers to ensure the semantics are correct.

File: src/stage1/ir.cpp (hunk 3 above)
Comment (m-radomski) [reply src/stage1/ir.cpp:3342]: If you look at rounding, `rem` and `mod` do them differently.

File: lib/std/special/compiler_rt/floatfmodl.zig (hunk 2 above)
Comment (m-radomski) [reply lib/std/special/compiler_rt/floatfmodl.zig:7]: This function relay's on integer overflows, enabling it causes the ubsan to trap.

File: test/behavior/math.zig [hunk 4]
```diff
@@ -791,6 +791,33 @@ fn remdiv(comptime T: type) !void {
...
+    comptime try frem(f32);
+    comptime try frem(f64);
+    comptime try frem(f128);
+    try frem(f16);
+    try frem(f32);
+    try frem(f64);
+    try frem(f128);
+}
+
+fn frem(comptime T: type) !void {
```
Comment (matu3ba) [inline test/behavior/math.zig:805]: Add the corresponding tests + those with different results than `mod` to these tests.

File: src/stage1/ir.cpp (hunk 3 above)
Comment (matu3ba) [reply src/stage1/ir.cpp:3342]: ah, sounds good. mark this as resolved, if you want.

File: lib/std/special/compiler_rt/floatfmodl.zig (hunk 2 above)
Comment (matu3ba) [reply lib/std/special/compiler_rt/floatfmodl.zig:7]: Zig has explicit integer overflow (called [wrapping overflow](https://ziglang.org/documentation/master/#Wrapping-Operations)) with `+%`, `-%`, `*%`. Why does this not work?
Silent overflow can lead to overwrite of other register values, which in turn leads to incorrect results.

Comment (m-radomski) [reply lib/std/special/compiler_rt/floatfmodl.zig]: Done in 201761b6

File: lib/std/special/compiler_rt/floatfmodl_test.zig (hunk 1 above)
Comment (m-radomski) [reply lib/std/special/compiler_rt/floatfmodl_test.zig:25]: Done in 834c9e25

File: test/behavior/math.zig (hunk 4 above)
Comment (m-radomski) [reply test/behavior/math.zig:805]: Done in 1dd7e624

**Review by matu3ba — CHANGES_REQUESTED** [review]
`(a*b)/(a*b)` is `0/0` for `b=0`. please explain.

File: lib/std/special/compiler_rt/floatfmodl.zig [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    var bmod = b;
+    const aPtr = @ptrCast(*ReinterpretUnion, &amod);
+    const bPtr = @ptrCast(*ReinterpretUnion, &bmod);
+
+    const signA = aPtr.parts.exp_and_sign & 0x8000;
+    var expA = @intCast(i32, (aPtr.parts.exp_and_sign & 0x7fff));
+    var expB = bPtr.parts.exp_and_sign & 0x7fff;
+
+    if (b == 0 or std.math.isNan(b) or expA == 0x7fff) {
+        return (a * b) / (a * b);
```
[the remark this answers is not in the document]
Comment (m-radomski) [reply lib/std/special/compiler_rt/floatfmodl.zig:56]: Done in [b87f788](https://github.com/ziglang/zig/pull/10833/commits/b87f7885b8a67e973f0a2d221455c2f016933dba)

**Comment by m-radomski** [comment]
Drafting this change because:
- stage2 does not handle `extern struct` yet
- Windows and MacOS do not link, because `fmodl` has different names on those systems

CI is showing passing I presume because all the calculations are happening at comptime.

**Pull request merged**